Introduce integration mode of sanitycheck

nashif opened the pull request:
his adds a new option, integration which is used to list platforms to use in integration testing (CI) and avoids whitelisting platforms. This will enable running those tests on real hardware if the --integration option is omitted.
--integration option is now not default, in the future this will be the default and will build/test a minimal set, running sanitycheck with more coverage will be possible use another option.

oh, one more thing in this PR: use native_posix exclusively where it makes sense and avoid massive build/running trivial stuff that is being tested elsewhere anyways..

resolves #25104.

nashif committed:
sanitycheck: realign schema

Just a cleanup and alignment of the yaml file.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: introduce integration testing

Use --integration testing for CI to limit execution and builds on
platforms that actually provide most of the coverage instead of blindly
building/running on all available platforms.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: philosophers: enable for integration testing

Enable sample for integration.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: cmsis_dsp: use integration keyword

Enable sample for integration.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: hello_world: filter for integration

Enable sample for integration.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: cpp_synchronization: filter for integration

Enable sample for integration.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
ci: enable sanitycheck for integration

Enable sample for integration.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: dac: remove whitelist

We already depend on dac.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: benchmarks: filter for integration

Enable for --integration option with sanitycheck.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: tinycrypt: do not whitelist

use integration filtering...

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: bluetooth: cleanup prj*.conf

Remove commented out options

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
drivers: arcv2_irq_unit: remove include

This include is not needed, code fails to build when configuration is
changed..

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
power: remove unused inclusion of soc.h

Inclusion of soc.h not needed in this file

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: device: remove whitelisting

Remove whitelisting and use --integration instead.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests: shell: remove filter

This filter is bogus, we already add it into the prj.conf

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
tests/samples: enable for integration testing

Limit execution in CI when using --integration option of sanitycheck.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
samples: pi: add integration support

Limit execution in CI when using --integration option of sanitycheck.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

stephanosio commented on tests/benchmarks/mbedtls/testcase.yaml:
```diff
@@ -3,4 +3,6 @@ common:
 tests:
   benchmark.crypto.mbedtls:
-    platform_whitelist: qemu_x86 frdm_k64f sam_e70_xplained nrf52840dk_nrf52840
+    filter: CONFIG_TEST_RANDOM_GENERATOR
+    integration:
+      - qemu_x86
```
Is there a reason `frdm_k64f sam_e70_xplained nrf52840dk_nrf52840` are removed from the integration testing list?

I can see the value of running these "integration tests" on them, since `mbedtls` is a crypto library and should ideally be tested/validated on every major supported platform if not all.

nashif replied:
> I can see the value of running these "integration tests" on them, since `mbedtls` is a crypto library and should ideally be tested/validated on every major supported platform if not all.

validated how? We are just building those tests right now, integration testing limits that to only 1 runnable platform in CI. Removing the whitelisting here enables this on every single platform we have including those that were whitelisted and those that were not.

stephanosio replied:
> validated how? We are just building those tests right now, integration testing limits that to only 1 runnable platform in CI. Removing the whitelisting here enables this on every single platform we have including those that were whitelisted and those that were not.

@nashif I was under the impression that the `integration:` platform list was migrated with the future support for hardware CI testing in mind (i.e. I see other commits in this PR that left `frdm_k64f` and `sam_e70_xplained` in the `integration:` list).

PerMac replied:
If I understood correctly from the discussion about the `integration` list it supposes to contain only emulated platforms where tests/samples will be run by zephyr's CI. I've seen `fedm_k64f` and `sam_e70_xplained ` added to the integration list in cmsis_dsp tests but (again if I got it correct from @nashif description) it was due to a 'hassle' with being able to add the `integration` list and all the samples/tests will have to be still reevaluated to correctly handle the `integration`.

ceolin replied:
That is not clear for me. What is the criteria used to pick where to do the integration test ?

Related to this specific test,it should be tested in different platforms, mbedtls has some integration with platform that should be tested as well, for example random number generation. This benchmarking test is exercising several functions and would be nice to have them running in as much targets as possible.

stephanosio commented on samples/bluetooth/mesh_demo/prj_nrf51_blenano.conf:
```diff
@@ -18,16 +17,3 @@ CONFIG_BT_MESH_MODEL_GROUP_COUNT=2
 CONFIG_BT_MESH_PB_ADV=n
 CONFIG_BT_MESH_CFG_CLI=y
-
-#CONFIG_BT_MESH_LOW_POWER=y
-#CONFIG_BT_MESH_SELF_TEST=y
-#CONFIG_BT_MESH_IV_UPDATE_TEST=y
-#CONFIG_BT_MESH_DEBUG=y
-#CONFIG_BT_MESH_DEBUG_NET=y
-#CONFIG_BT_MESH_DEBUG_TRANS=y
-#CONFIG_BT_MESH_DEBUG_PROV=y
-#CONFIG_BT_MESH_DEBUG_BEACON=y
-#CONFIG_BT_MESH_DEBUG_CRYPTO=y
-#CONFIG_BT_MESH_DEBUG_ADV=y
-#CONFIG_BT_MESH_DEBUG_ACCESS=y
-#CONFIG_BT_DEBUG_LOG=y
```
I wonder if these were intentionally left here so as to remind users of the existence of these potentially useful/relevant options (after all, this is a "sample").

nashif replied:
those commented options bitrot really fast and become irrelevant if they are not tested. If there is anything to remind the user, it should be part of the documentation. I argue this was just left over from someone debugging the sample.

PerMac commented on scripts/sanity_chk/sanitylib.py:
```diff
@@ -1208,12 +1208,6 @@ def get_test(self, name, valid_keys):
-                    self.filename,
-                    "Unknown config key '%s' in definition for '%s'" %
-                    (k, name))
-
```
What is the reason behind this change? Is it because 'integration' is not in the list of valid_keys? Why not adding 'integration' there and keeping this check in place?

nashif replied:
this is already validated by the yaml schema checks, it is just an old check we had that is duplicating what yaml verification does now.

galak reviewed (dismissed):
Need to update doc/guides/test/sanitycheck.rst.

I wish we had a better name than `integration`, best I can come up with is `ci`.

andyross commented:
> I wish we had a better name than integration

Interestingly, the most common term of art in the industry for this kind of simplest-possible-but-still-comprehensive test to check for obvious bugs before merge is a "sanity check".  Though I'm reliably told that in the modern world that language is frowned on as not particularly inclusive.

Seriously, I've been staying out of this one for obvious reasons but it all looks good to me.

galak commented:
> Interestingly, the most common term of art in the industry for this kind of simplest-possible-but-still-comprehensive test to check for obvious bugs before merge is a "sanity check". Though I'm reliably told that in the modern world that language is frowned on as not particularly inclusive.
> 
> Seriously, I've been staying out of this one for obvious reasons but it all looks good to me.

I wasnt thinking of it in those terms, just that integration didn't spring to mind the meaning of build/run this test for CI purposes.

stephanosio commented:
> I wasnt thinking of it in those terms, just that integration didn't spring to mind the meaning of build/run this test for CI purposes.

We have `platform_whitelist` to list the platforms that are supported by a test/sample, I think something like `ci_whitelist` would be more intuitive.

PerMac commented:
Some ideas for the name (I tried to avoid uninclusive language).
CI_smoketest, CI_BVT (build verification test), CI_verification, CI_smokelist, CI_scope. The CI prefix is optional, however, I think using it clearly shows the purpose of the list.

nashif commented:
> I wish we had a better name than `integration`, best I can come up with is `ci`.

what is the i in ci? :)

galak commented:
> what is the i in ci? :)

I know :)

ceolin commented:
> > I wish we had a better name than `integration`, best I can come up with is `ci`.
> 
> what is the i in ci? :)

I have no problem with the name, the only thing I can think against is that this may bring some confusion with unit and integration tests, that in this cases refers to tests different components integration.

ceolin commented:
This option is not a replacement for `platform_whitelist` right ? In this case why `platform_whitelist` as removed in some commits ?

PerMac commented:
> This option is not a replacement for `platform_whitelist` right ? In this case why `platform_whitelist` as removed in some commits ?

I believe the reason for this PR is the fact that whitelisting became an issue blocking proper on-target testing. If a whitelist exists it prevents the test from being run on any other platform, which is especially a problem in the context of test on the physical boards. Whitelisting ended up as a simple way to provide minimum test coverage (e.g. one writes a test based on a platform on his/her desk and adds the platform to the whitelist, which prevents the test from being run on any other platform) and also a list to minimize the scope of the CI testing to a limited subset of platforms. I believe the trend is to depreciate whitelisting and decouple 2 issues: the scope for the CI tests is given by the 'integration' list and the filtering of the tests runnability on a given platform will be based on dts compatibility.

nashif committed:
doc: add documentation about --integration option

Document the --integration option.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

carlescufi commented:
Also not a fan of `integration` but cannot come up with something better either. Perhaps `ci` would be shorter and easier to understand?

Also, this closes #25104 right @nashif?

nashif commented:
> Also not a fan of `integration` but cannot come up with something better either. Perhaps `ci` would be shorter and easier to understand?

Again, ci is 'continuous integration', this is 'integration', so not sure what the issue is here.. This can also be run out of CI if someone wants a faster run locally, so it is not limited to ci. 

and yes, it fixes  #25104.

galak commented on doc/guides/test/sanitycheck.rst:
```diff
@@ -303,7 +303,16 @@ arch_exclude: <list of arches, such as x86, arm, arc>
-    Set of platforms that this test case should only be run for.
+    Set of platforms that this test case should only be run for. Do not use
+    this option to limit testing or building in CI due to time or resource
+    constraints, this option should only be used of the test or sample can
```
`used of` I think is suppose to be `used if`.

galak commented on doc/guides/test/sanitycheck.rst:
```diff
@@ -303,7 +303,16 @@ arch_exclude: <list of arches, such as x86, arm, arc>
+    constraints, this option should only be used of the test or sample can
+    only be run on the whitelisted platform and nothing else.
+
+integration: <YML list of platforms/boards>
```
Maybe `platform_integration_list` or `integration_platforms` instead of `integration`?

nashif replied:
Another proposal: drop all of this and do it completely differently...

basically the option would be --mode with a list of modes that are supported, for example:
``--mode [daily|pull-request|default|...]```


now, in the testcase yaml file this can be triggered with:

```

 test:
   mode:
     pull-request:
       - <platform>
       - <platform>
     daily:
       - all
```
Still need to figure out the syntax and allow flexibility here.

Existing `build_on_all` would fold into that and removed as a standalone option for example.

nashif replied:
ok, sticking with original proposal with a name change, now using integration_platforms.

galak reviewed (dismissed):
* Minor typo in docs.
* Idea of using a slightly more descriptive term in YAML.

erwango commented on tests/drivers/dac/dac_loopback/testcase.yaml:
```diff
@@ -1,6 +1,5 @@
+  tags: dac drivers
 tests:
   drivers.dac:
-    platform_whitelist: frdm_k64f nucleo_l073rz nucleo_l152re twr_ke18f
```
This won't be sufficient.
In `tests/drivers/dac/dac_loopback/src/test_dac.c` there is some board specific configuration code, that ends up in:
```
#else
#error "Unsupported board."
#endif
```
So I think that with current implementation, the platform_withelist should stand.

nashif replied:
this depends on dac already, so that should be fine.

erwango replied:
So each board declaring dac support will have to go and edit this test_dac.c, but maybe at the end someone will try to find a smarter test implementation... ok

LeiW000 reviewed (approved):
It looks good to me.

nashif committed:
boards: qemu_x86_coverage: limit to core tests

Limit to kernel tests, this should not try and run every sample or test
we have as it only verifies the coverage feature, but not collecting
coverage.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
